Test internal builds

forabi opened the pull request:
Closes hollowverse/hollowverse#167

forabi committed:
Rebase on master

forabi commented:
We need to decide what to do when a requested test environment is not found. Should we soft fail to master or should we show an error page?

forabi committed:
Pass port in dev command

forabi committed:
Clear production env cookie in test environments

forabi committed:
Skip production middleware if it is a test environment

forabi committed:
Fix cookie parsing

forabi committed:
Wrap prod middleware in try/catch

forabi committed:
Lock EB API version

wholesomedev commented on src/testInternalBuilds.ts:
```diff
@@ -0,0 +1,39 @@
+
+testInternalBuilds.use(async (req, res, next) => {
+  try {
+    const requestedEnvName: string | undefined = req.query.env;
```
I think the query param should be called `branch`

forabi replied:
Renamed

wholesomedev reviewed (commented):
I'm not sure using separate middleware functions is the best approach here. I'm also not sure about the naming.

With regard to the naming, the word `test` might invoke certain connotations that don't exactly describe what this code does. (I know the original purpose of Release Manager is to ultimately do testing, my so-called Real Testing™, but this is just one piece of the puzzle so I don't think we should apply the name "test" to it)

The purpose of this code is to route to a branch: `master`, `beta`, or `deploy-*`. So, I'm thinking the names can be: `defaultTrafficSplitter` and `branchPreviewSplitter` or `defaultEnvironmentRouter` or `branchPreviewRouter` or something like that.

With regard to the use of middleware functions: you will always want to use either the branch preview splitter or the default splitter, never both. So, I think we should use one middleware that checks if the query contains the `branch` query param or not. If it does, get the environment URL from the branch preview splitter. If it doesn't, get the environment URL from the default splitter.

Just a suggestion.

forabi committed:
Merge branch 'master' into test-internal-builds

forabi committed:
environmentsByUrl => urlsByEnvironment

forabi committed:
Simplify code handling traffic splitting and previewing branches

forabi committed:
Merge branch 'test-internal-builds' of github.com:forabi/release-manager into test-internal-builds

forabi commented:
@msafih I've greatly simplified the code handling both routing paths. It's no longer using separate middlewares, but it keeps the logic for fetching the environment details separate for maintainability.

forabi committed:
Fix clearCookie, do not throw on query string error

forabi committed:
Only pass env cookie value to cookie handler

forabi committed:
Rename target to endpoint

forabi committed:
Extract env cookie name to variable

forabi committed:
Rename env getters, support cookie for test branches

forabi commented:
@msafih I've also changed the code so that now we only have to set the query string one time and all subsequent pages will be served from the test branch. This is done via another cookie called `branch` that takes precedence over the `env` cookie, but only if the test branch actually exists, so when the branch environment is deleted, the code soft-fails to a random production environment.

forabi committed:
Reduce branch cookie maxAge to 2 hours

forabi committed:
Rename endpoint to url

forabi committed:
Add comments for env getters

wholesomedev commented on src/releaseManager.ts:
```diff
@@ -1,58 +1,57 @@
 
-const getEnvName = createEnvNameGenerator(weightsByEnvironment);
+process.on('unhandledRejection', () => {
+  setIsHealthy(false);
```
We'll eventually want to log these somewhere, like SumoLogic?

forabi replied:
@msafih I'm thinking we should extract the log wrapper to a separate package so we can use it in all applications and take advantage of the typed events.

msafih commented:
>@msafih I've also changed the code so that now we only have to set the query string one time and all subsequent pages will be served from the test branch. This is done via another cookie called `branch` that takes precedence over the `env` cookie, but only if the test branch actually exists, so when the branch environment is deleted, the code soft-fails to a random production environment.
>
>But this also means that as long as the test environment is still up, we will be served that environment until the cookie expires. We might forget that we are actually using a test environment. We will need to remove the cookie manually to return to a production version. I've reduced the `branch` cookie `maxAge` to 2 hours.

Cool. 

When we hit a page from the browser, how are we going to know which environment it's being served from? Will this information be in the header or somewhere else?

forabi committed:
Merge branch 'master' into test-internal-builds

forabi commented:
> When we hit a page from the browser, how are we going to know which environment it's being served from? Will this information be in an HTTP header or some where else?

Both cookies `branch` and `env` can be viewed in Chrome DevTools in the Application tab.

msafih commented:
@forabi Okay, but this is not very convenient and it's something that **we set**. Not something that the server responds with. The question is: how do we know what the server responded with?

forabi commented:
@msafih How about we set a custom header with the endpoint as the value
just before we call `proxy.web`?

On Sat, Sep 23, 2017, 7:27 PM msafih <notifications@github.com> wrote:

> @forabi <https://github.com/forabi> Okay, but this is not very convenient
> and it's something that *we set*. Not something that the server responds
> with. The question is: how do we know what the server responded with?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hollowverse/release-manager/pull/5#issuecomment-331648059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA90GcYfK1RR4ITBWaxMJxJKc6QQfSEcks5slTFjgaJpZM4PfFkb>
> .
>

forabi committed:
Fix tslint paths and enable type checking rules

msafih commented:
@forabi That should be fine if the `endpoint` is a user-friendly string from which it is easy tell the branch name. Otherwise, it would be better to have the branch name in the in the header.

forabi committed:
Remove 'hollowverse-' prefix from cookies

forabi committed:
Add X-Hollowverse-Release-Manager-Target

This custom header shows the final environment name as resolved by Release Manager.

forabi commented:
@msafih I've added a new header, `X-Hollowverse-Release-Manager` which includes the final branch name as resolved by Release Manager. https://github.com/hollowverse/hollowverse/pull/197 also adds a similar header to version server to serve as the source of truth in case something goes wrong in the routing code.

forabi committed:
Rename header and add branch info

forabi commented:
@msafih I added another header, `X-Hollowverse-Requested-Environment`, which is set when requesting a branch with the query string param.

So if we request `https://hollowverse.com/?branch=test`:

* `X-Hollowverse-Requested-Environment` will be `test`.
* `X-Hollowverse-Resolved-Environment` will be `test` if the environment `test` actually exists. Otherwise it will be a randomly assigned environment like `master` or `beta`
* `X-Hollowverse-Actual-Environment` (https://github.com/hollowverse/hollowverse/pull/197): should match the previous header if release manager logic is good.

Without `?branch=test`, `X-Hollowverse-Requested-Environment` will not be set.

msafih commented:
👍

forabi commented:
@msafih I think this PR is ready

wholesomedev commented on src/releaseManager.ts:
```diff
@@ -1,58 +1,62 @@
+        maxAge: 2 * 60 * 60 * 1000,
+      });
+    } else {
+      res.clearCookie(branchPreviewCookieName);
```
When will this get executed? If there's no `env` for the requested branch, the cookie won't be set to begin with, so why do you clear it if it doesn't get set?

forabi replied:
What if it was set and the test environment was deleted afterwards? If we don't remove the cookie we will have to send a request to EB API on every request.

wholesomedev commented on src/trafficSplitter/getRandomEnvName.ts:
```diff
@@ -12,7 +12,7 @@ const chance = new Chance();
  * one set of 5 sessions may differ from the next 5 sessions.
  */
-export function* createEnvNameGenerator(
+export function* createRandomEnvNameGenerator(
```
This isn't really "random"

forabi replied:
The choice will be random within the constraints of the defined weights. Although every 5 requests will include 4 master and 1 beta, the order of these is random because we are using `chance.pickone`.
